File: concurrent-programming-with-monitors/FlagDelay.java
/* 
 * Concurrent programming with monitors (Java)
 * by Project Nayuki, 2024. Public domain.
 * https://www.nayuki.io/page/concurrent-programming-with-monitors
 */

import java.util.Random;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;


public final class FlagDelay {
	
	// Customizable parameters
	private static final int NUM_WAITERS = 4;
	private static final int MAX_CHECK_DELAY_MS = 1_000;
	private static final int MAX_SET_DELAY_MS = 10_000;
	
	
	// State variables
	private static final long startTime = System.currentTimeMillis();
	private static Lock lock = new ReentrantLock();
	private static boolean flag = false;
	
	
	public static void main(String[] args) throws InterruptedException {
		for (int i = 0; i < NUM_WAITERS; i++)
			new Thread(FlagDelay::worker, "worker" + i).start();
		
		int duration = rand.nextInt(MAX_SET_DELAY_MS);
		timestampedPrintf("Sleep %d ms%n", duration);
		Thread.sleep(duration);
		
		lock.lock();
		try {
			timestampedPrintf("Set true%n");
			flag = true;
		} finally {
			lock.unlock();
		}
	}
	
	
	private static void worker() {
		try {
			while (true) {
				lock.lock();
				try {
					if (flag)
						break;
				} finally {
					lock.unlock();
				}
				
				int duration = rand.nextInt(MAX_CHECK_DELAY_MS);
				timestampedPrintf("Get false; sleep %d ms%n", duration);
				Thread.sleep(duration);
			}
		} catch (InterruptedException e) {
			throw new RuntimeException(e);
		}
		timestampedPrintf("Get true%n");
	}
	
	
	private static void timestampedPrintf(String format, Object... args) {
		long time = System.currentTimeMillis() - startTime;
		var newArgs = new Object[args.length + 2];
		newArgs[0] = time;
		newArgs[1] = Thread.currentThread().getName();
		System.arraycopy(args, 0, newArgs, 2, args.length);
		synchronized(FlagDelay.class) {
			System.out.printf("[%5d ms] %s | " + format, newArgs);
		}
	}
	
	
	private static Random rand = new Random();
	
}

File: concurrent-programming-with-monitors/flag-delay.py
# 
# Concurrent programming with monitors (Python)
# by Project Nayuki, 2024. Public domain.
# https://www.nayuki.io/page/concurrent-programming-with-monitors
# 

import random, threading, time


# Customizable parameters
_NUM_WAITERS: int = 4
_MAX_CHECK_DELAY_S: float = 1.0
_MAX_SET_DELAY_S: float = 10.0


# State variables
_start_time: float = time.time()
_lock: threading.Lock = threading.Lock()
_flag: bool = False


def main() -> None:
	for i in range(_NUM_WAITERS):
		threading.Thread(name=f"Worker{i}", target=_worker).start()
	
	duration: float = random.uniform(0.0, _MAX_SET_DELAY_S)
	_timestamped_print(f"Sleep {round(duration * 1000)} ms")
	time.sleep(duration)
	
	_timestamped_print("Set true")
	with _lock:
		global _flag
		_flag = True


def _worker() -> None:
	while True:
		with _lock:
			if _flag:
				break
		
		duration: float = random.uniform(0.0, _MAX_CHECK_DELAY_S)
		_timestamped_print(f"Get false; sleep {round(duration * 1000)} ms")
		time.sleep(duration)
	
	_timestamped_print("Get true")


_print_lock: threading.Lock = threading.Lock()

def _timestamped_print(s: str) -> None:
	relativetime: float = time.time() - _start_time
	s = f"[{int(relativetime * 1000):5d} ms] {threading.current_thread().name} | {s}"
	with _print_lock:
		print(s)


if __name__ == "__main__":
	main()
